### ISSUE TITLE
Incorrect `selectAll` typing

### ISSUE THREAD

**Issue by romeerez** [kysely-org/kysely#1059]
Hello dear maintainers, I found an incorrect typing in the [Not null](https://kysely.dev/docs/examples/select/not-null) docs:

```ts
     jsonObjectFrom(
       eb.selectFrom('pet')
         // the problem is in this selectAll:
         .selectAll()
         .limit(1)
         .whereRef('person.id', '=', 'pet.owner_id')
     ).$notNull().as('pet')
```

SQL is correctly selecting * from the pet table, so only the pet columns are actually selected, but the typing combines columns of all tables.

```ts
// pet doesn't have a first_name, but no complaint from TS
persons[0]?.pet?.first_name
```

**Comment by igalklebanov** [comment]
Hey 👋

Thank you for submitting this crucial issue!

As a workaround, please scope the `selectAll` to get the expected results:

```ts
selectAll('pet')
```

**Comment by igalklebanov** [comment]
To solve this we need to find a way to track the "external" query context (tables that exist in the main query's `from` clause) aside from the "local" query context (tables that exist in the subquery's `from` clause). As `ExpressionBuilder` passes the "external" query context tables `TB` currently represents the "total" query context ("external" and "local") that is causing this issue.

My intuition points at an additional (4th) generic @ `SelectQueryBuilder` representing the tables from the "external" query context - what `ExpressionBuilder` knows about.
The simplest solution, which should only affect `SelectQueryBuild.selectAll()` (`select *`) would be to exclude "external" query context tables from `TB`, while all other methods keep using `TB` as-is.

We are usually conservative when it comes to adding more generics. However, output type correctness is important, and these helpers are frequently used, so IMHO this time it's worth it.

**Comment by koskimas** [comment]
>We are usually conservative when it comes to adding more generics. However, output type correctness is important, and these helpers are frequently used, so IMHO this time it's worth it.
I strongly disagree with this one. I definitely don't think we should add a fourth generic just for this case as it's easily fixed by using `selectAll('pet')`. Fixing this would be complicated and it'd only solve this one particular case.

**Comment by romeerez** [comment]
As an option, `selectAll()` without arguments could be deprecated, so users won't accidentally select everything from all the joined tables, and it would solve the typing issue.

**Comment by koskimas** [comment]
> As an option, `selectAll()` without arguments could be deprecated, so users won't accidentally select everything from all the joined tables, and it would solve the typing issue.

`select *` could and should almost always be replaced by `select table.*` for sure, but there might be some legitimate use cases for `select *`. Preventing people from from writing it is too opinionated IMO.

**Comment by hughgardiner** [comment]
Perhaps this belongs in a separate issue, but I'm curious how feasible it is to have the typescript response types accurately reflect the correct types based on the order of `.selectAll('table')` statements.

For example:
```typescript
const rows = await db
  .selectFrom("person")
  .leftJoin("pet", "pet.person_id", "person.person_id")
  .selectAll("person")
  .selectAll("pet")
  .execute();
```
Because this generates 
```sql
select person.*, pet.*
```
the `pet` fields take priority in the output, which makes `person_id` possibly null given the field `left join`'d on has the same name on both tables.

However, the generated response type says `person_id` is not nullable. if you reverse the order, the expected outcome of `person_id` being non-nullable is accurate as the `person.person_id` field will always be present
